### PR TITLE
Add Categorical support

### PR DESCRIPTION
**Pull request by cpcloud**

### CODE REVIEW

File: datashape/coretypes.py [hunk 1]
```diff
@@ -1031,6 +1035,38 @@ def __getitem__(self, key):
...
+    __slots__ = 'type', 'categories'
+    cls = MEASURE
+
+    def __init__(self, categories, type=None):
+        self.categories = tuple(categories)
+        self.type = type or datashape.discover(self.categories).measure
+
+    def __repr__(self):
+        return '%s[%s]' % (type(self).__name__.lower(),
+                           pformat(self.categories))
```
Comment (mrocklin) [inline datashape/coretypes.py]: This repr might get out of hand

Comment (cpcloud) [inline datashape/coretypes.py]: What would you suggest here? I can limit the length that's printed with `pformat`

Comment (cpcloud) [inline datashape/coretypes.py]: Hm, well that turns out to be not true

**Comment by jcrist** [comment]
What still needs to be done on this?

**Comment by cpcloud** [comment]
@llllllllll @jcrist any more thoughts on this before merge?

File: datashape/coretypes.py [hunk 2]
```diff
@@ -9,8 +9,11 @@
 
 import ctypes
 import operator
+
 from math import ceil
 
+import datashape
```
Comment (jcrist) [inline datashape/coretypes.py]: datashape imports itself, probably by accident?

Comment (cpcloud) [inline datashape/coretypes.py]: no, that's being used elsewhere

File: datashape/coretypes.py [hunk 3]
```diff
@@ -1005,6 +1008,58 @@ def __getitem__(self, key):
     def __str__(self):
         return pprint(self)
 
+category_predicates = [
+    (lambda x: 0 < x <= 1 << 8, np.uint8),
```
Comment (cpcloud) [inline datashape/coretypes.py]: these are unnecessary now, since i'm not actually doing the categorization

File: datashape/coretypes.py [hunk 4]
```diff
@@ -1005,6 +1008,58 @@ def __getitem__(self, key):
...
+    (lambda x: 1 << 8 < x <= 1 << 16, np.uint16),
+    (lambda x: 1 << 16 < x <= 1 << 32, np.uint32),
+    (lambda x: 1 << 32 < x <= 1 << 64, np.uint64)
+]
+
+
+def dtype_for_length(n):
```
Comment (cpcloud) [inline datashape/coretypes.py]: this can also be gutted

File: datashape/coretypes.py (hunk 2 above)
Comment (jcrist) [inline datashape/coretypes.py]: Ah, I see it's to deal with circular imports. nvm.

**Comment by jcrist** [comment]
Looks good to me. Thanks for finishing this up.

**Comment by cpcloud** [comment]
Ok, merging

**Pull request merged**